### PR TITLE
Support dual firmware (aka A/B) devices where rootfs_data are discreete partitions

### PR DESCRIPTION
**Pull request by Lanchon**
hi @hauke @dangowrt @robimarko

Many devices support dual firmwares (`kermel`/`kernel_1`, `rootfs`/`rootfs_1`). However openwrt only supports dual firmwares on devices for which the overlay is stoted in the trailing part of the active rootfs partition. Devices with discreet overlay partitions (`rootfs_data`/`rootfs_data_1`) are not supported.

This change allows the bootloader to choose which `rootfs_data` partition to use, just like it could always choose which `rootfs`: via a kernel command line parameter.

For example, uboot can now specify `fstools_overlay_name=rootfs_data` or `fstools_overlay_name=rootfs_data_1`. This works for block devices, UBI and MTD, as the system delegates partition name resolution to the usual fstools driver stack.

For completeness, an `fstools_overlay_fstype=ext4`/`f2fs`/`auto` parameter is also included. This selects which filesystem to use when it is time to format the overlay anew.

thank you!

-----

this change has been tested to work properly in spectrum sax1v1k.

### CODE REVIEW

**Comment by Lanchon** [comment]
hi @hauke, sorry to bother you.

openwrt does not support dual-firmware in devices with discreet `rootfs_data` partitions. this is a serious limitation but easy to fix. who should i bring in to review this change? thank you!

**Comment by Lanchon** [comment]
tangential note:

i must admit that (in addition to these changes) i'd also like seeing a u-boot env var that takes precedence over `fstools_overlay_fstype` to select the overlay filesystem.

in time i've heard many people wanting ext4 in the overlay in the forums, and i see no reason why we shouldn't accommodate them. a simple uboot var could be set to solve the issue for those who care. i could also do that change if desired.

**Comment by Lanchon** [comment]
@dangowrt so there's no interest in supporting dual firmware?

**Comment by dangowrt** [comment]
I think there could be interest in supporting existing vendor bootloaders which offer A/B dual-boot feature. As OpenWrt typically uses "the remaining space" on the flash for a read/write overlay, when implemented from scratch it should be done in such way that there still is only one `rootfs_data` partition which gets wiped once switching from A to B

**Comment by Lanchon** [comment]
@dangowrt 

> which is what this PR is about, ie. supporting the existing A/B on NBG7815, right?

no, i don't have that device, this is about spectrum sax1v1k. the relevant partitions of this device are:
```
  20           68130          330273   128.0 MiB   FFFF  rootfs
  21          330274          346657   8.0 MiB     FFFF  0:WIFIFW
  22          346658          608801   128.0 MiB   FFFF  rootfs_1
  23          608802          625185   8.0 MiB     FFFF  0:WIFIFW_1
  24          625186         1673761   512.0 MiB   FFFF  rootfs_data
  25         1673762         2722337   512.0 MiB   FFFF  rootfs_data_1

  32         5164578         5172769   4.0 MiB     FFFF  rsvd_1
  33         5172770         5180961   4.0 MiB     FFFF  rsvd_2
  34         5180962         5185057   2.0 MiB     FFFF  rsvd_3
  35         5185058         5217825   16.0 MiB    FFFF  rsvd_4
  36         5217826         5283361   32.0 MiB    FFFF  rsvd_5
  37         5283362         5414437   64.0 MiB    FFFF  rsvd_6

  38         5414438        15204321   4.7 GiB     FFFF  user_data
```

this device is has secure boot enabled and a locked down bootloader. the bootloader cannot be interrupted. initial access is gained by hacking the firmware userland. then uboot vars are programmed that will patch uboot in memory before loading an unsigned kernel.

the situation was very precarious because a borked system would have become an unrecoverable brick. i coded a [solution](https://forum.openwrt.org/t/spectrum-sax1v1k-askey-rt5010w-openwrt-support/149923/216) for this, allowing breaking the boot process if you have serial access, an also externally triggering a recovery (initramfs) OS at any time without serial access.

on top of that, i also wanted to add the possibility of completely rolling back a sysupgrade externally without serial access, but i'm not going to do that if both systems share rootfs_data. i think it just doesn't make sense to do that.

the two rootfs_data partitions are there, but i can't use them because openwrt is missing this capability. the partitions cannot be merged or renamed, as [in my experience](https://github.com/Lanchon/REPIT/issues/55) most qualcomm systems from the last decade with secure boot enable sign the GPT, so any change results in a brick. (not that i'd want to merge anything: i think 512 MB rootfs_data is more than enough, and IMHO the best and safest way to leverage an 8GB emmc like this one in openwrt is to use a big partition (eg: the existing user_data with 4.7 GiB) as extra storage that survives sysupgrades.)

### actual status of official devices

> As OpenWrt typically uses "the remaining space" on the flash for a read/write overlay, when implemented from scratch it should be done in such way that there still is only one rootfs_data partition

i respectfully disagree. this might have been the case in 128 MB devices, but there is no use in having 3.5 GB or 7.5 GB rootfs_data partitions these days.

more importantly, this is not how actual official openwrt devices are implemented. many of [these devices](https://docs.openwrt.melmac.net/luci-app-advanced-reboot/) have separate rootfs_data partitions for A and B (placing rootfs_data after rootfs via loop) in both nand and emmc storages.

#### example: whw03v2
512 MB NAND device where rootfs_data could be 450 MB but is only 120 MB. here are the partitions:
```
[    1.328823] 0x000000700000-0x00000a800000 : "kernel"     : 161 MiB (6 MiB kernel + rootfs)
[    1.490871] 0x000000d00000-0x00000a800000 : "rootfs"     : 155 MiB
[    1.647219] 0x00000a800000-0x000014900000 : "alt_kernel" : 161 MiB (6 MiB kernel + rootfs)
[    1.809863] 0x00000ae00000-0x000014900000 : "alt_rootfs" : 155 MiB
[    1.959552] 0x000014900000-0x000014b00000 : "sysdiag"    :   2 MiB (unused)
[    1.961963] 0x000014b00000-0x000020000000 : "syscfg"     : 181 MiB (stock state, ext4)
```
- a 330 MB rootfs_data could have been made out of the last 3 partitions, and shared by A and B, but instead each system has its 120 MB trailing area of rootfs. reasons are compatibility with stock, ease of rolling back to stock, etc.
- a 450 MB rootfs_data could have been made by moving alt_kernel (it is only referenced by a uboot var) or by using the mtdconcat driver to add rootfs to the 330 MB partition.

(btw i proposed [solutions](https://github.com/openwrt/openwrt/issues/15439) to this, but they are not relevant here.)

#### example: whw03v1
a 4 GB emmc GPT device where rootfs_data could be 3.5 GB but is less than 100 MB. this also has separate rootfs_data for A and B (as loop devices after rootfs). for this device i provided a [solution](https://openwrt.org/toh/linksys/whw03_v1#repartitioning): optional steps to repartition the device to use 2 x 512 MB rootfs_data (i don't see the point in using more) and the rest of the emmc goes to an extra partition that survives sysupgrades, which you can use if you happen to need it.

there are many more examples in the list linked above.

### my dual boot experience

after using proper dual boot devices (with dual rootfs_data) i'm really not going back. situations in which a misconfiguration would have required ripping off the device from its network and access it failsafe in isolation happened many times. this is inconvenient enough, but sometimes the devices where remote, which would have meant somebody knowledgeable would have needed to commute. thanks to proper dual boot, these situations just required a special power up sequence to roll back changes, and the rest was handled via VPN.

### reasons for this change

- some devices, such as the sax1v1k, just have the extra rootfs_data partition there. and touching the GPT at all very probably results in a brick, so it can't be merged. so there is no point in not allowing people to use it.
- on nand devices, which tend to be smaller, repartitioning is an easy choice: just boot the new kernel and the repartition is effectively done. on GPT devices, repartition is much more work, and typically not done in standard installs. also going back to stock from a repartitioned device is more challenging. many people using GPT devices end up using a very small portion of their storage because they don't want to repartition. a way out of this mess is using an existing partition not named rootfs_data for that purpose, which this change allows. so this change is not only for A/B systems, but also for single boot systems with difficult stock GPTs.
- on GPT devices with secure boot enabled, this change becomes essential. you cannot modify the GPT (including names), so how do you handle the case of a device that has a good partition candidate for rootfs_data but is not named exactly that? (answer: you don't, and cram everything into a small rootfs.) this is more general than just the A/B case.
- even if dual rootfs_data were not the officially preferred install method, there is no reason not to allow advance users to customize their installs. a user could choose to split his "standard" big rootfs_data into 2 smaller partitions, modify the uboot env, and be happy with his dual boot, dual data system. GPT systems allow more end-user customization than nand systems; there is no reason not to support that.

### conclusion: better support for GPT

so in summary, openwrt partition handling was designed for nand flash, where the device port author could freely name freely defined partitions, so names could be hard-coded in code as they are. in GPT devices names preexist, and they are either very cumbersome to change (repartition during install) to downright impossible (signed GPTs). in light of this, a way to configure names is needed to better support GPT systems.

configuration via the kernel command line is for sure needed. it might be desirable to have other avenues for that, idk, but this is a start.

**Comment by Lanchon** [comment]
@dangowrt

could you maybe review this or suggest someone who should?

**Comment by Lanchon** [comment]
hi @robimarko,

what would be the best way to have a discussion about this?

maybe someone in the core project is interested, but i don't know how to find that out

**Comment by Lanchon** [comment]
friends...

so i waited some 8 months with my Spectrum SAX1V1K (Askey RT5010W) opened up so i can write an A/B-compatible uboot config and upgraded A/B sysupgrade routines for it. this work was blocked by openwrt not supporting A/B boot on emmc devices, which this PR intended to cure.

i am closing up my router now. i no longer have the time to do A/B sysupgrade for it.

this change is still valuable for other people interested in A/B sysupgrade, so i suggest you keep it open, but no longer for me.

thanks!

**Review by dangowrt — APPROVED** [review]

**Comment by Ansuel** [comment]
@Lanchon it's sad that you took this decision and I'm checking this only now... the code is so simple it's I can approve it right away with some changes...

Any chance you can give this a second chance with my guidance?

**Comment by Lanchon** [comment]
hi @Ansuel 

i have a device, the Spectrum SAX1V1K, that is secure boot-enabled but hacked, and would hard brick if an upgrade went south. so i made [bootloader scripts](https://github.com/Lanchon/openwrt-sax1v1k) for it that can be triggered to recover without even serial access, check them out if you want.

the device is partitioned well, and has dual kernel, rootfs, and rootfs_data partitions, which would allow for A/B updates, except openwrt currently can handle that. (btw, repartitioning to get a larger rootfs_data is not an option, as it would probably hard brick because qulacomm's android secure boot stack typically verifies the signature of the GPT itself.)

> it's sad that you took this decision and I'm checking this only now...

i had additions to the BL scripts and openwrt thought out to support safe A/B updates (without ever touching uboot's environment in the process), but now i don't have the time to rethink and implement all that. that's what i meant earlier.

> Any chance you can give this a second chance with my guidance?

but sure, we can still work on these patches.

thanks!

**Comment by Lanchon** [comment]
@Ansuel 

the justifications for these changes can be found [here](https://github.com/openwrt/fstools/pull/9#issuecomment-2436271542), especially starting on section "my dual boot experience".

**Comment by Ansuel** [comment]
My only complain for this is fstools_overlay_fstype... Why we need this? 

First and second commit are ok... The only change needed is to keep following the pattern with 
```
type
function_name
```
reducing the patch delta only to the specific changes.

**Comment by Lanchon** [comment]
@Ansuel

thank you for helping!
i have a couple of question...

---

> The only change needed is to keep following the pattern

[this file](https://github.com/openwrt/fstools/blob/master/libfstools/common.c) is a mixture of `int f()` and `int<CR>f()` conventions.

**Q1) should i fix it or not?**

---

`fstools_overlay_name` can be used to implement A/B boot.

but it has another use...

imagine a GPT device that comes with a large partition named `user_data`. this partition would be ideal to store the overlay, but i can't be done because the GPT is signed and the partition cannot be renamed.

this case and be solved using `fstools_overlay_name` in the dts:

```
	chosen {
		bootargs = "fstools_overlay_name=user_data";
	};
```

**Q2) but what happens if both `bootargs` and the command line passed from u-boot define `fstools_overlay_name` at the same time?**

**Q3) and if `bootargs-append` is used instead?**

---

> `fstools_overlay_fstype`... Why we need this?

it is less important, but i included it for completeness.

on GPT devices, fstools always chooses to format overlay in ext4 or f2fs according to the size of the overlay block device. it uses a non-configurable cutoff size. this behavior cannot be changed. so openwrt builds for GPT devices must always include the ext4 and f2fs user space utilities and kernel modules, even if only one type of file system will ever be used.

**case 1:** a device has a small block device and ext4 will always be chosen. the maintainer would want to save space and not include f2fs user space utilities and kernel module. he can do it safely with:

```
	chosen {
		bootargs = "fstools_overlay_type=ext4";
	};
```

**case 2:** some users don't trust f2fs much and prefer to use ext4. especially on routers, because shutdown is usually unclean. (also, f2fs is only better for systems that write a lot to the disk, and openwrt typically doesn't.) `fstools_overlay_type` lets advanced users choose their preferred file system by changing the boot environment.

**Q4) should we include `fstools_overlay_type` or not? it is not very important.**

---

**Q5) what is the best name for `fstools_overlay_name`**?

- `fstools_overlay_name`
- `fstools_overlay`
- `overlay_name`

---

let me know what you think and ill fix the PR.
thank you!

**Comment by Ansuel** [comment]
@Lanchon no sorry! Just having too much browser tabs open and forgot to reply! Are you still here?

**Comment by Lanchon** [comment]
im sailing in seychelles, but we brought a starlink, so yes go ahead :)

**Comment by Ansuel** [comment]
yes please!

**Review by dangowrt — APPROVED** [review]

**Comment by dangowrt** [comment]
Sorry that this took so long, from my side mostly because there isn't a very good solution to this problem, and using cmdline parameters felt a bit like a work-around for something which should be addressed differently. However, in absence of a better solution (which will anyway require bootloader integration and not be suitable for all existing devices out there), let's go ahead and apply this, it add two long-awaited features and shouldn't hurt anyone (apart from the negligible size-increase of libfstools)

**Comment by Lanchon** [comment]
@dangowrt @Ansuel 

thank you so much Dan!!!

i still think there should be a value for `fstools_overlay_name` that basically says "do not match any partition and use a loop device after rootfs instead".

(why? please see https://github.com/openwrt/fstools/issues/10. TLDR: fstools has a bug that makes it ignore `rootfs_data` on some devices with dual storage. on those devices, we need to manually specify using a loop for overlay **before** fixing the bug in fstools, else there will be data loss.)

right now we can use `fstools_overlay_name=WhATEvA_637#` with a string that does not match any partition, but that is extremely hacky.

i can add a trivial PR to support:
- either the empty string `fstools_overlay_name=` (this *might* work now but is brittle and difficult to test: the code will just pass the empty string to the `find` function of each of the configured storage drivers, so every driver must handle the empty string correcly, not to mention some partitions not having a name is a thing)
- or a specific magic label like `fstools_overlay_name=-` or `=[loop]` or `=@none`
- or another param like `fstools_overlay_loop=1`

IMHO there should be a canonical way to express this. what do you think?

**Comment by Ansuel** [comment]
force_loop ? My only concern for this was the overload of variables and I would have loved to handle them in the format fstools.overlay.force_loop.

After all it's all matching specific string so having something like empty string i see it as fragile (I have seen lots of downstream project where they love to specify every variable but just empty... just for the sake of having gigantic long lines.)
the [loop] or `@none`  might be an idea but more to embed that in a dedicated variable...

But my question is... since we have fstools_overlay_fstype can't we just set that to loop and it will force loop mounting?

That is used only on creating new filesystem but can't we also make it to force one? Example change the logic and check... 

we want an ext4. If it's present mount, if it's not format. If we found something else, force it. (and in the case of force if that wouldn't work then fail.... in the case of loop it will always work)

@dangowrt what do you think? I feel forcing a fstype might be ideal for this case.

**Comment by Lanchon** [comment]
> But my question is... since we have fstools_overlay_fstype can't we just set that to loop and it will force loop mounting?

> I feel forcing a fstype might be ideal for this case.

that would be a good solution if loop had a specific unique FS associated with it. but it doesn't... they are orthogonal settings and they even apply to different devices: _name to all, and _type only to block devices. (and on block devices, the FS used for loop depends on its size unless specified.) i don't really like this idea, _type is not the right mechanism IMHO.

> I would have loved to handle them in the format fstools.overlay.force_loop.

ok, so this PR had been open for a long time and i didnt quite remember what i saw when i made it, but there is already:
- [fstools_partname_fallback_scan](https://github.com/search?q=repo%3Aopenwrt%2Ffstools%20fstools_partname_fallback_scan&type=code): used in [chromium](https://github.com/openwrt/openwrt/blob/7ba45de8275845bd1d85d0d9ff9d4662610ce844/target/linux/ipq806x/image/chromium.mk#L31)
- [fstools_ignore_partname](https://github.com/search?q=repo%3Aopenwrt%2Ffstools%20fstools_ignore_partname&type=code): used in [2 devices](https://github.com/search?q=repo%3Aopenwrt%2Fopenwrt%20fstools_ignore_partname&type=code)

`fstools_ignore_partname` maybe works at the wrong level (it affects all searches instead of particular searches such as overlay) and only works for the partname driver, but i think it sort of does the force loop thing.

> having something like empty string i see it as fragile (I have seen lots of downstream project where they love to specify every variable but just empty...

well right now `fstools_overlay_name=` WILL break a device expected to use `rootfs_data`. the problem is that it won't do anything useful reliably (besides acting different than option not present). if those downstream projects want to specify every variable, right now they have to use `fstools_overlay_name=rootfs_data`, not `fstools_overlay_name=`.

but we need to make `fstools_overlay_name=` do something reliably: either force a loop in all cases (my preference, and what it mostly does now) or have the system ignore the param. but either solution needs a change.

my position:

- preferably have `fstools_overlay_name=` reliably force a loop
- else have `fstools_overlay_name=` totally ignored, and add `fstools_overlay_loop=1`
